docs/root/configuration/http_filters/jwt_authn_filter.rst

```diff
@@ -0,0 +1,184 @@
+
+  Authorization: Bearer <token>
+
+If fails to extract a JWT from above header, then check query parameter key *access_token* as in this example::
+
+  /path?access_token=<JWT>
+
+In the :ref:`filter config <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.JwtAuthentication>`, *providers* is a map, to map *provider_name* to a :ref:`JwtProvider <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.JwtProvider>`. The *provider_name* must be unique, it is referred in the `JwtRequirement <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.JwtRequirement>` in its *provider_name* field.
+
+**Important note** for *remote_jwks*, a **jwks_cluster** cluster is required for Envoy to talk to a JWKS server. Due to this, `OpenID connection discovery <https://openid.net/specs/openid-connect-discovery-1_0.html>`_ is not supported since the URL to fetch JWKS is in the response of the discovery. It is not easy to setup a cluster config for a dynamic URL.
+
+Remote JWKS config example
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. code-block:: yaml
+
+  providers:
+    provider_name1:
+      issuer: https://example.com
+      audiences:
+      - bookstore_android.apps.googleusercontent.com
+      - bookstore_web.apps.googleusercontent.com
+      remote_jwks:
+        http_uri:
+          uri: https://example.com/jwks.json
+          cluster: example_jwks_cluster
+        cache_duration:
+          seconds: 300
+
+Above example fetches JWSK from a remote server with URL https://example.com/jwks.json. The token will be extracted from the default extract locations. The token will not be forwarded to upstream. JWT payload will not be added to the request header.
+
+Following cluster **example_jwks_cluster** is needed to fetch JWKS.
+
+.. code-block:: yaml
+
+  cluster:
+    name: example_jwks_cluster
+    type: STRICT_DNS
+    hosts:
+      socket_address:
+        address: example.com
+        portValue: 80
+
+
+Inline JWKS config example
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Another config example using inline JWKS:
+
+.. code-block:: yaml
+
+  providers:
+    provider_name2:
+      issuer: https://example2.com
+      local_jwks:
+        inline_string: PUBLIC-KEY
+      from_headers:
+      - name: jwt-assertion
+      forward: true
+      forward_payload_header: x-jwt-payload
+
+Above example uses config inline string to specify JWKS. The JWT token will be extracted from HTTP headers as::
+
+     jwt-assertion: <JWT>.
+
+JWT payload will be added to the request header as following format::
+
+    x-jwt-payload: base64_encoded(jwt_payload_in_JSON)
+
+RequirementRule
+~~~~~~~~~~~~~~~
+
+:ref:`RequirementRule <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.RequirementRule>` has two fields:
+
+* Field *match* specifies how a request can be matched; e.g. by HTTP headers, or by query parameters, or by path prefixes.
+* Field *requires* specifies the JWT requirement, e.g. which provider is required.
+
+**Important notes:**
```

Comment:
Can you use Sphinx RST directives for this? http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#directives